Promote security policies to stable

aryan16 opened the pull request:
**Graduation of Security policies(authz and reqAuthn) from V1beta to V1. Added a separate entry for PeerAuthn in feature.yaml.**

**Design Doc** - https://docs.google.com/document/d/1h3QKiIRuJUIX0XYTgg7pZD1k_moa2S-oazmGDde_W3w/edit#heading=h.5mzkohub2lz4

**Work Tracking Doc** - https://docs.google.com/document/d/1GN8eyTcwnGGA89efFtfkrjmoczFw0jPnHpf5pcPAoAg/edit#heading=h.32ht3qk19wnh

**L7 Deny Rules on TCP Ports Issue**: 
Updated the doc and added a warning if users have L7 deny rule which blocks all the TCP traffic unexpectedly -
docs - https://github.com/istio/istio.io/pull/12178, https://github.com/istio/api/pull/2561
L7 deny policy validation - https://github.com/istio/istio/pull/41802

howardjohn reviewed (commented):
My main concern which I wanted opinions from other TOC members is the relationship to gateway-api. Presumably long term there will be either a common API or at least a common way to attach authz policies to workloads.

So on one hand, I don't want to promote to stable when we plan to change it.

On the other hand, we can't make breaking changes to these APIs _anyways_, so it is practically stable. This is the same approach `Ingress` took in Kubernetes, with the same problem.

We also (so far) have been taking the first approach for networking APIs. However, the gateway-api is _much_ further along for those APIs.

Thoughts? Also adding @craigbox

craigbox commented:
I would like to see all the existing Istio APIs promoted to v1, with the same logic that applied to Ingress in Kubernetes.  It sends a strong signal that you can keep using your supported resources as we start a migration to a new set of APIs.

That said, they didn't just rev the version -- they had to do a non-zero amount of work to get to an acceptable v1, and it's very easy for me to spend other people's money.

costinm commented:
I would wait until K8S Gateway/Mesh WG makes some progress on authz. I
think the starting point can be the Istio RBAC, but likely there will be
some change.

But I don't mind that much if we just move what we have to v1 - there will
be some work to update the CRDs, and users migrating their resources to
istio.v1 and
later to gateway API RBAC extension.


On Mon, Jul 18, 2022 at 11:27 AM Aryan Gupta ***@***.***>
wrote:

> I agree with Craig, we already have our apis stable. There is no harm in
> promoting them as it would definitely increase the confidence among our
> users.
> @howardjohn <https://github.com/howardjohn> what’s your take on this ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/enhancements/pull/122#issuecomment-1188064721>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2QLTFVOJTQS335U5ULVUWOXTANCNFSM53IHWCHQ>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

linsun commented:
Makes sense to me, i assume this doesn't include peer authentication policy?

aryan16 commented:
> Makes sense to me, i assume this doesn't include peer authentication policy?

yeah, no plans of promoting peer auth policy

lei-tang commented:
@aryan16 Thank you for updating the PR based on what we discussed in the TOC meeting today!

ericvn commented on features.yaml:
```diff
@@ -208,12 +208,20 @@ features:
+    level:
+      checklist: ""
+      maturity: Stable
+      nextExpectedPromotion: ""
```
If we are promoting this to stable, we should really have a filled out checklist. It looks like we could start with what's in this PR: https://github.com/istio/enhancements/pull/42.

aryan16 replied:
somehow there is no `jwt_authentication.md` in `/features`, should I create a new one for jwt authentication promotion ?

lei-tang replied:
jwt_authentication.md is a new file in https://github.com/istio/enhancements/pull/42, which has not been merged. I think you can create a jwt_authentication.md (or an equivalent) as the checklist of RequestAuthentication.

lei-tang replied:
Meanwhile, you may want to refer to the checklist of another stable feature when filling the checklist of RequestAuthentication.

costinm commented:
I did a bit more research on this and played with some configs - I think
it's reasonable to move to v1, no need to delay based on K8S Gateway
policies.

The main difference is that L4 and L7 are mixed - in Gateway they are
usually separated, and the mix of L4 and L7 makes it hard to have
implementations
where L4 policy is applied at the L4 layer more efficiently.  Knowing if a
policy should be applied at L7 is based on the presence of L7
attributes/conditions
and/or the port protocol.

I don't know if while moving to v1 we can make any changes - like making it
explicit when an authorization policy should be applied at L7, or adding
a stronger language on how an app should identify that a port is explicitly
L7.

However - this is probably low priority, I would expect Gateway to define a
separate policy for L4 and L7 ( since they also split routing, and given
their
attachment policy ) - also because other vendors may have other opinions
and structure may need to align with the policy, so there is very little
chance for the new API to be identical with what we have. By moving to v1
we make it clear the current API is frozen and long-term supported in
context of Istio APIs/sidecar model, while support for Gateway API will use
whatever comes out.

On Mon, Aug 8, 2022 at 1:31 PM lei-tang ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In features.yaml
> <https://github.com/istio/enhancements/pull/122#discussion_r940638528>:
>
> >        nextExpectedPromotion: ""
>      area: Security and policy enforcement
>    - name: "End User (JWT) Authentication"
>      id: "security.jwt_authentication"
> -    link: "/docs/concepts/security/#authentication"
> +    link: "/docs/reference/config/security/request_authentication/"
> +    level:
> +      checklist: ""
> +      maturity: Stable
> +      nextExpectedPromotion: ""
>
> Meanwhile, you may want to refer to the checklist of another stable
> feature when filling the checklist of RequestAuthentication.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/enhancements/pull/122#discussion_r940638528>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2QZUUFZ5TRRW6ZCO5TVYFVDTANCNFSM53IHWCHQ>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

lei-tang commented on features/jwt_authentication.md:
```diff
@@ -0,0 +1,195 @@
+- [x] Test all the possible fields from CRD to imporve the test coverage. 
+- [x] Fix issues in multicluster scenarios.
+
+**Performance**
```
Consider adding a comment/plan for the Performance section.

aryan16 replied:
I think we have it already - https://istio.io/latest/blog/2020/large-scale-security-policy-performance-tests/, https://github.com/istio/tools/tree/master/perf/stability/control-plane. I don't think we need anything else for this. Will update the file with completed status

ericvn replied:
I think this is fine to mark complete.

costinm commented:
Also please add tests for unexpected configs - for example policies with L7
attributes on L4 ports, verify how L4 policies apply on both L4 and L7
ports.

On Mon, Aug 8, 2022 at 4:23 PM lei-tang ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In features/jwt_authentication.md
> <https://github.com/istio/enhancements/pull/122#discussion_r940743455>:
>
> > +- [x] The supportability review panel has reviewed promotion of the feature.
> +- [x] The TOC has reviewed and approved promotion of the feature as part of the
> +	road map for a release.
> +
> +---
> +
> +## Stable
> +
> +### Requirements:
> +
> +**Tests**
> +
> +- [x] Test all the possible fields from CRD to imporve the test coverage.
> +- [x] Fix issues in multicluster scenarios.
> +
> +**Performance**
>
> Consider adding a comment/plan for the Performance section.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/enhancements/pull/122#pullrequestreview-1065901602>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUR2XPQ4VP2TGI523DEWDVYGJIJANCNFSM53IHWCHQ>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

aryan16 commented:
Updated the doc and added a warning if users have L7 deny rule which blocks all the TCP traffic unexpectedly - 
docs - https://github.com/istio/istio.io/pull/12178
L7 deny policy validation - https://github.com/istio/istio/pull/41802

We are good to go with this promotion, please review this PR again.

aryan16 commented:
PTAL : @howardjohn @costinm @louiscryan @ericvn @linsun

ericvn commented on features/authorization.md:
```diff
@@ -190,15 +191,22 @@ Istio Authorization provides access control for workloads in the mesh at the nam
-- [ ] The supportability review panel has reviewed promotion of the feature.  
+- [X] The appropriate work group(s) have reviewed and approved promotion of the feature.
+- [X] The supportability review panel has reviewed promotion of the feature.  
 - [ ] The TOC has reviewed and approved promotion of the feature as part of the
```
I think you can mark this complete as the TOC will have approved promotion.

aryan16 replied:
updated, thanks

ericvn reviewed (commented):
LGTM with a couple of fields to mark complete. Thanks for adding the features.md file.
